src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -432,14 +515,33 @@ public string ServerName
 
         public override byte[] GetValue()
         {
-            byte[] tmpValue = null;
-            if (ServerName != null)
+            int sizeEstimate = 10 + 2 * (ServerName?.Length ?? 0);
+            AsnWriter writer = new(AsnEncodingRules.BER, sizeEstimate);
+
+            /* This is as laid out in MS-ADTS, 3.1.1.3.4.1.16.
+             * VerifyNameRequestValue ::= SEQUENCE {
+             *                              Flags       INTEGER,
+             *                              ServerName  OCTET STRING }
+             */
+            using (AsnWriter.Scope outerSequence = writer.PushSequence())
             {
-                UnicodeEncoding unicode = new UnicodeEncoding();
-                tmpValue = unicode.GetBytes(ServerName);
+                writer.WriteInteger(Flag);
+
+                if (!string.IsNullOrEmpty(ServerName))
+                {
+                    int serverNameLength = Encoding.Unicode.GetByteCount(ServerName);
+                    Span<byte> tmpValue = serverNameLength < 256 ? stackalloc byte[serverNameLength] : new byte[serverNameLength];
```

Comment:
nit:
```suggestion
                    int serverNameLength = Encoding.Unicode.GetByteCount(ServerName);
                    Span<byte> tmpValue = serverNameLength <= 256 ? stackalloc byte[serverNameLength] : new byte[serverNameLength];
```

also I wonder if this is a place for constant sized stackalloc? (IIRC it has bit better codegen)
e.g. 
```cs
const int MaxStackalloc = 256;
.. = serverNameLength <= MaxStackalloc ? 
    stackalloc byte[MaxStackalloc].Slice(0, serverNameLength) : new byte[serverNameLength];
```